Add BreakIteratorBoundaryScanner support

shaie opened the pull request:
This commit adds a boundary_scanner property to the search highlight
request (defaults to "simple", per current behavior) so the user can
specify "break_iterator" and its type ("sentence", "word", "line" and
"character"). Also adds "boundary_scanner_locale" to define which one
should be used when highlighting the text.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:
```diff
@@ -119,10 +127,12 @@ protected AbstractHighlighterBuilder(StreamInput in) throws IOException {
         order(in.readOptionalWriteable(Order::readFromStream));
         highlightFilter(in.readOptionalBoolean());
         forceSource(in.readOptionalBoolean());
+        boundaryScannerType(in.readOptionalWriteable(BoundaryScannerType::readFromStream));
```
Since this PR can go to 5.x can you add a check on the version, like this:
````` 
 if (in.getVersion().onOrAfter(Version.V_5_4_0_UNRELEASED)) { 
    boundaryScannerType(in.readOptionalWriteable(BoundaryScannerType::readFromStream));
}
`````
This will ensure that a node in 5.4 can safely send request to a 5.3 node.

shaie replied:
Nice, I didn't see this pattern in the files I changed, so had no idea :). I will do that in all 4 places you commented on.

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:
```diff
@@ -119,10 +127,12 @@ protected AbstractHighlighterBuilder(StreamInput in) throws IOException {
         if (in.readBoolean()) {
             boundaryChars(in.readString().toCharArray());
         }
+        boundaryScannerLocale(in.readOptionalString());
```
same here

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:
```diff
@@ -150,12 +160,18 @@ public final void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalWriteable(order);
         out.writeOptionalBoolean(highlightFilter);
         out.writeOptionalBoolean(forceSource);
+        out.writeOptionalWriteable(boundaryScannerType);
```
and here ...

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:
```diff
@@ -150,12 +160,18 @@ public final void writeTo(StreamOutput out) throws IOException {
+        out.writeBoolean(hasBoundaryScannerLocale);
+        if (hasBoundaryScannerLocale) {
+            out.writeString(boundaryScannerLocale.toLanguageTag());
+        }
```
and the last one ;)

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java:
```diff
@@ -522,4 +525,33 @@ public String toString() {
     }
+
+    public static enum BoundaryScannerType implements Writeable {
+        SIMPLE, BREAK_ITERATOR;
```
BREAK_ITERATOR is not reflecting what this type is doing. Maybe just SENTENCE ? Or SENTENCE_BREAK ?

shaie replied:
This is actually exactly what it means. As I wrote in the PR comment, I wanted to get initial feedback. Now that I see you're OK with the general approach, I want to add a "break_iterator_type" property which defaults to "SENTENCE" but also supports "WORD", "CHARACTER" and "LINE".

jimczi replied:
Got it thanks. Though we could avoid creating another option and merge the logic in a single one ? The boundary scanner type could be any of:
* simple
* sentence
* word
The fact that we use a break iterator for sentence and word is just an implementation detail ? 
Although "CHARACTER" seems weird, should we really expose it ?

shaie replied:
I don't think we _have_ to expose all options, was just thinking "if Lucene supports it ..". But having said that, from an API perspective this may be weird, so I'll take your proposal and add "boundary_scanner" with values "simple", "sentence" and "word". I'll do that.

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java:
```diff
@@ -522,4 +525,33 @@ public String toString() {
+            if (boundaryScannerType.toUpperCase(Locale.ROOT).equals(BREAK_ITERATOR.name())) {
+                return BoundaryScannerType.BREAK_ITERATOR;
+            }
+            return SIMPLE;
```
I think it's too lenient, we should fail if the type is unknown.

shaie replied:
I followed the same approach as in `Order` though I agree with you. Just thought this is how ES prefers to do things. I can simplify the method to use `valueOf(boundaryScannerType.toUpperCase(Locale.ROOT))`

jimczi replied:
++

shaie commented:
@jimczi I believe I've addressed all your comments. Hope I addressed the BW ones as you intended!

shaie commented:
@jimczi I only now noticed your comment about adding something to the docs. Can you point me to where that is?

jimczi commented:
Thanks @shaie ! You can check the current docs for the FVH here: https://github.com/elastic/elasticsearch/blob/master/docs/reference/search/request/highlighting.asciidoc#fast-vector-highlighter
I'll review your changes shortly.

shaie commented:
@jimczi I updated the docs and also renamed `SIMPLE` to `CHARS`, which I think makes more sense?

jimczi commented:
@elasticmachine please test it

shaie commented:
@jimczi, the build failed but seems weird exception, something about Git plugin ..

jimczi commented:
Yes it's just a problem with our CI. I'll test locally :(

shaie commented:
K thanks @jimczi. Let me know if there are any issues, though I did run some tests locally and they passed (ones I thought are related to the change).

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/AbstractHighlighterBuilder.java:
```diff
@@ -150,12 +165,22 @@ public final void writeTo(StreamOutput out) throws IOException {
+            boolean hasBoundaryScannerLocale = boundaryScannerLocale != null;
+            out.writeBoolean(hasBoundaryScannerLocale);
+            if (hasBoundaryScannerLocale) {
+                out.writeString(boundaryScannerLocale.toLanguageTag());
```
out.writeOptionalString() since you use in.readOptionalString in the readFrom ?

shaie replied:
No, notice that since `boundaryScannerLocale` is a `Locale`, I first write a boolean, and then write the String if it's not null. If I wanted to use `writeOptionalString` I would need to do something like:

`out.writeOptionalString(boundaryScannerLocale == null ? null : boundaryScannerLocale.toLanguageTag())`

Do you think that's preferred?

jimczi replied:
++

shaie replied:
Just for reference @jimczi, I borrowed this pattern from the code just above those lines which handles `boundaryChars`:

```
        boolean hasBounaryChars = boundaryChars != null;
        out.writeBoolean(hasBounaryChars);
        if (hasBounaryChars) {
            out.writeString(String.valueOf(boundaryChars));
        }
```

So before I change the code I wrote, does this change your opinion? not that I mind changing the code, just prefer the code to be consistent with the rest.

jimczi replied:
I don't mind as long as we use `writeString/readString` and `writeOptionalString/readOptionalString` consistently. 
So you can maybe just change the `readFrom` to explicitly use readBoolean.

shaie replied:
OK done.

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilder.java:
```diff
@@ -522,4 +525,30 @@ public String toString() {
         }
     }
+
+    public static enum BoundaryScannerType implements Writeable {
```
this makes our code checker unhappy since enum are always static in this context.

shaie replied:
Good point!

jimczi commented on core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightBuilderTests.java:
```diff
@@ -558,12 +560,23 @@ private static void setRandomCommonOptions(AbstractHighlighterBuilder highlightB
+                highlightBuilder.boundaryScannerType(randomFrom(BoundaryScannerType.values()));
+            } else {
+                // also test the string setter
+                highlightBuilder.boundaryScannerType(randomFrom(BoundaryScannerType.values().toString()));
```
this fails the build, the `toString` is not correctly applied

shaie replied:
Right, fixed.

jimczi commented on core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/SearchContextHighlight.java:
```diff
@@ -324,6 +348,9 @@ Builder merge(FieldOptions globalOptions) {
                 }
+                if (fieldOptions.boundaryScannerType == null) {
+                    fieldOptions.boundaryScannerType = globalOptions.boundaryScannerType;
+                }
```
The merge of the boundaryScannerLocale is missing

shaie replied:
Sorry about that. Fixed.

shaie commented:
I'm running `gradle clean check` and will push a new commit to the PR once it finishes. Sorry for the noise @jimczi

shaie commented:
@jimczi I pushed a commit which addresses all your comments and all tests pass. Actually, when running `gradle clean check` only one test failed: `org.elasticsearch.client.transport.NodeDisconnectIT.testNotifyOnDisconnect`, but it doesn't look related to this PR and also the failures looked like:

```
 1> Caused by: org.elasticsearch.transport.NodeNotConnectedException: [node_s0][127.0.0.1:9500] Node not connected
  1>    at org.elasticsearch.transport.TcpTransport.getConnection(TcpTransport.java:594) ~[main/:?]
  1>    at org.elasticsearch.transport.TcpTransport.getConnection(TcpTransport.java:114) ~[main/:?]
  1>    at org.elasticsearch.transport.TransportService.getConnection(TransportService.java:509) ~[main/:?]
  1>    at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:485) ~[main/:?]
  1>    ... 19 more
```

I can rerun, but if you want to start the tests too, wanted to let you know.

jimczi commented:
Thanks @shaie !
I'll merge to 5.x now.

shaie commented:
Thanks @jimczi!
